codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -0,0 +1,266 @@
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    private void consumeBytes(int streamId, int bytes) {
+        try {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            http2Handler.connection().local().flowController()
+                        .consumeBytes(stream, bytes);
+        } catch (Throwable t) {
+            exceptionCaught(ctx, t);
+        }
+    }
+
+    private void writeGoAwayFrame(Http2GoAwayFrame frame, ChannelPromise promise) {
+        int lastStreamId;
+        if (frame.lastStreamId() == null) {
+            lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+            + frame.extraStreamIds() * 2;
+        } else {
+            lastStreamId = frame.lastStreamId();
+        }
+
+        http2Handler.goAway(
+                http2HandlerCtx, lastStreamId, frame.errorCode(), frame.content().retain(), promise);
+    }
+
+    private void writeStreamFrame(Http2StreamFrame frame, ChannelPromise promise) {
+        int streamId = frame.streamId();
+        if (frame instanceof Http2DataFrame) {
+            Http2DataFrame dataFrame = (Http2DataFrame) frame;
+            http2Handler.encoder().writeData(http2HandlerCtx, streamId, dataFrame.content().retain(),
+                                             dataFrame.padding(), dataFrame.isEndStream(), promise);
+        } else if (frame instanceof Http2HeadersFrame) {
+            Http2HeadersFrame headerFrame = (Http2HeadersFrame) frame;
+            http2Handler.encoder().writeHeaders(
+                    http2HandlerCtx, streamId, headerFrame.headers(), headerFrame.padding(), headerFrame.isEndStream(),
+                    promise);
+        } else if (frame instanceof Http2ResetFrame) {
+            Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
+            http2Handler.resetStream(http2HandlerCtx, streamId, rstFrame.errorCode(), promise);
+        } else {
+            throw new UnsupportedMessageTypeException(frame);
+        }
+    }
+
+    private final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(stream.id()));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            ctx.fireUserEventTriggered(new Http2StreamClosedEvent(stream.id()));
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(errorCode, debugData);
+            goAway.setLastStreamId(lastStreamId);
+
+            // TODO: Should we also set extraStreamIds?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)